🌱   refactor: standardize machine filter functions and improve testing

CecileRobertMichon opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:  The util, controllers, kcp internal packages had overlapping functions that fetch or filter many machines. Some functions use the clusterv1.MachineList type, others use []*clusterv1.Machine, others use collections.Machines. This PR standardizes the use of filter functions in util/collections across the codebase.

(Issue came out this discussion: #2037 (comment))

In addition, I found a few "bugs" in tests of those functions - tests that weren't actually testing what they said they were testing so I fixed those.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2062

CecileRobertMichon commented on controllers/machine_controller_test.go:
```diff
@@ -1377,12 +1303,20 @@ func TestIsDeleteNodeAllowed(t *testing.T) {
 		{
-			name:    "no control plane members",
-			cluster: &clusterv1.Cluster{},
+			name: "no control plane members",
```
this tests was not fully correct as the machines were being excluded from the filter by not having the right cluster name label and/or namespace. Added all the required object metadata so it's actually checking that the described missing piece (eg. No nodeRef) is what causes the error.

CecileRobertMichon commented on controlplane/kubeadm/controllers/status_test.go:
```diff
@@ -46,6 +46,10 @@ func TestKubeadmControlPlaneReconciler_updateStatusNoMachines(t *testing.T) {
 	}
 
 	kcp := &controlplanev1.KubeadmControlPlane{
+		TypeMeta: metav1.TypeMeta{
```
this one got me debugging the test for a while 😬 

the test was actually a no-op because of https://github.com/kubernetes-sigs/cluster-api/pull/4207/files#diff-a6639fd91ffb9cff50ef1f3af76ec2371aa5c41e0d28b98b651cb9193eeaab62L52

Since the fake management cluster `Management ` was nil, it wasn't actually listing machines and filtering them, it was just returning ALL machines that were passed into managementCluster.Machines. When I changed it to actually using the filter, it was failing because the Machines weren't "owned" by KCP since the KCP TypeMeta was missing here but not in the owner ref, it was comparing empty string to the ref's Kind, Group, and Version.

CecileRobertMichon commented on util/util.go:
```diff
@@ -100,80 +90,6 @@ func Ordinalize(n int) string {
-
-// ModifyImageRepository takes an imageName (e.g., repository/image:tag), and returns an image name with updated repository
-// Deprecated: Please use the functions in util/container
-func ModifyImageRepository(imageName, repositoryName string) (string, error) {
```
there's some overlap with #4005 here, I don't mind rebasing on top of it once it merges.

This is ready for review in the meantime though.

/hold

CecileRobertMichon replied:
rebased on top of #4078

/hold cancel

CecileRobertMichon commented:
```
sigs.k8s.io/cluster-api/util
  Incompatible changes:
  - HasPausedAnnotation: removed
  - ImageTagIsValid: removed
  - IsPaused: removed
  - MachineListFormatDeprecationMessage: removed
  - ModifyImageRepository: removed
  - ModifyImageTag: removed
  - PointsTo: removed
  - SemverToOCIImageTag: removed
  - WatchOnClusterPaused: removed
  ```
  
  All of should be removed in https://github.com/kubernetes-sigs/cluster-api/pull/4078 already

CecileRobertMichon commented:
/milestone v0.4.0

vincepri commented on controllers/cluster_controller.go:
```diff
@@ -399,11 +400,12 @@ func (r *ClusterReconciler) listDescendants(ctx context.Context, cluster *cluste
 	// Split machines into control plane and worker machines so we make sure we delete control plane machines last
-	controlPlaneMachines, workerMachines := splitMachineList(&machines)
-	descendants.workerMachines = *workerMachines
+	controlPlaneMachines := collections.FromMachineList(&machines).Filter(collections.ControlPlaneMachines(cluster.Name))
+	workerMachines := collections.FromMachineList(&machines).Difference(controlPlaneMachines)
+	descendants.workerMachines = collections.ToMachineList(workerMachines)
```
```suggestion
        machineColleciton := collections.FromMachineList(&machines)
	controlPlaneMachines := machineCollection.Filter(collections.ControlPlaneMachines(cluster.Name))
	workerMachines := machineCollection.Difference(controlPlaneMachines)
	descendants.workerMachines = collections.ToMachineList(workerMachines)
```

vincepri commented on util/collections/machine_filters.go:
```diff
@@ -204,3 +215,22 @@ func MatchesKubernetesVersion(kubernetesVersion string) Func {
+
+// GetFilteredMachinesForCluster returns a list of machines that can be filtered or not.
+// If no filter is supplied then all machines associated with the target cluster are returned.
+func GetFilteredMachinesForCluster(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, filters ...Func) (Machines, error) {
```
Do we want to use a client within this package?

CecileRobertMichon replied:
it was already being used in the same func in the util package https://github.com/kubernetes-sigs/cluster-api/pull/4207/files#diff-a61d5e39f45fb5ed2dc1ef71fa708c8db480e9a396adb8a5d9d17b8e6ba1e186L129, this moves it to a more specific sub package. Would there be a better place to put it if we don't want any client usage in util/?

fabriziopandini replied:
What about moving to a helper.go file under utils/collections (It is a little bit confusing having this func into machine_filters)

CecileRobertMichon replied:
done

MarcelMue commented on controllers/machine_controller.go:
```diff
@@ -471,16 +457,16 @@ func (r *MachineReconciler) isDeleteNodeAllowed(ctx context.Context, cluster *cl
 	// number of remaining control plane members and whether or not this
 	// machine is one of them.
-	switch numControlPlaneMachines := len(util.GetControlPlaneMachines(machines)); {
+	switch numControlPlaneMachines := len(machines.Filter(collections.ControlPlaneMachines(cluster.Name))); {
```
Would it make sense to simplify this to an `if / else` statement? Would be a bit easier to read for me at least with only 2 outcomes.

MarcelMue commented on util/collections/machine_collection.go:
```diff
@@ -59,6 +59,15 @@ func FromMachineList(machineList *clusterv1.MachineList) Machines {
 	return ss
 }
 
+// ToMachineList creates a MachineList from the given Machines
```
```suggestion
// ToMachineList creates a MachineList from the given Machines.
```

CecileRobertMichon commented:
```
sigs.k8s.io/cluster-api/util
  Incompatible changes:
  - ImageTagIsValid: removed
  - MachineListFormatDeprecationMessage: removed
  - ModifyImageRepository: removed
  - ModifyImageTag: removed
``` 

All 4 are deprecated in alpha3: https://github.com/kubernetes-sigs/cluster-api/blob/release-0.3/util/util.go

CecileRobertMichon commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -270,7 +270,7 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	}
 
-	controlPlaneMachines, err := r.managementClusterUncached.GetMachinesForCluster(ctx, util.ObjectKey(cluster), collections.ControlPlaneMachines(cluster.Name))
+	controlPlaneMachines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster, collections.ControlPlaneMachines(cluster.Name))
```
let me look into this.

CecileRobertMichon replied:
changed it back to use r.managementClusterUncached and instead modified the implementation, PTAL

CecileRobertMichon commented:
@MarcelMue @fabriziopandini thanks for the reviews. Now that #4078 has merged I've rebased this PR and addressed your comments, PTAL.

fabriziopandini commented:
/lgtm

CecileRobertMichon commented:
/approve

vincepri reviewed (approved):
/approve
/lgtm
